File: library/src/main/java/org/incendo/interfaces/utilities/ThreadUtils.java
package org.incendo.interfaces.utilities;

import org.bukkit.Bukkit;
import org.bukkit.plugin.Plugin;
import org.incendo.interfaces.InterfacesListeners;

public final class ThreadUtils {

    private ThreadUtils() {
    }

    public static void runSync(final Runnable function) {
        if (Bukkit.isPrimaryThread()) {
            function.run();
            return;
        }

        Plugin plugin = InterfacesListeners.instance().plugin();

        if (plugin == null) {
            throw new IllegalStateException("Could not find plugin for class " + function.getClass().getName());
        }

        try {
            Bukkit.getScheduler().callSyncMethod(plugin, () -> {
                function.run();
                return null;
            }).get();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

}

File: next/src/main/java/org/incendo/interfaces/next/utilities/ThreadUtils.kt
package org.incendo.interfaces.next.view

import org.bukkit.Bukkit
import org.bukkit.plugin.java.PluginClassLoader

internal fun runSync(function: () -> Unit) {
    if (Bukkit.isPrimaryThread()) {
        function()
        return
    }

    val plugin = (function::class.java.classLoader as PluginClassLoader).plugin
    Bukkit.getScheduler().callSyncMethod(plugin, function)
}
